### PR TITLE
Add integration tests

### PR DESCRIPTION
**Pull request by qleisan**
Add initial pytest based integration tests

### CODE REVIEW

**Comment by rettichschnidi** [comment]
@qleisan Would you like some feedback already now? If not, please let us know once you do.

**Comment by qleisan** [comment]
Yes I do!
Some ongoing improvements 
I will refactor the helper/fixture code to have a base class and use inheritance (cleaner). Also I'm working on replacing the temporary "setformat" server CLI command with a command line argument for the client. 
e.g. lwm2mclient -f 11543 => the client registers with URI=...;ct=11543 and cause the server to use JSON when communicating with this client. However I just noticed some problems with this approach, registration doesn't work with all formats I want to use in testcases (investigating)

**Comment by qleisan** [comment]
Pushed "object oriented" refactoring of test helpers, thx to @mlasch for the suggestion. Odd bug (on my machine) one test executes twice (and fails second time)

File: tests/integration/test_bootstrap_interface.py
Comment (rettichschnidi) [inline tests/integration/test_bootstrap_interface.py:71]: Please remove references specific to your own workstation

File: tests/integration/conftest.py
Comment (rettichschnidi) [inline tests/integration/conftest.py:30]:
```suggestion
           self.__dumptext()
```

File: tests/integration/conftest.py
Comment (rettichschnidi) [inline tests/integration/conftest.py:70]: Please no commented or dead code.

File: tests/integration/conftest.py
Comment (rettichschnidi) [inline tests/integration/conftest.py:93]: Documentation should make clear what the difference to `lwm2mclient_boot` is.

File: tests/integration/conftest.py
Comment (rettichschnidi) [inline tests/integration/conftest.py:83]: Documentation should make clear what the difference to `lwm2mclient_boot` is.

File: tests/integration/test_bootstrap_interface.py
Comment (rettichschnidi) [inline tests/integration/test_bootstrap_interface.py:135]: I feed like many of this things should be implemented before we can merge this.

Also, this should be placed in a more prominent place (#586 to the rescue!) and somehow reflected in the commit message too.

File: tests/integration/test_bootstrap_interface.py
Comment (rettichschnidi) [inline tests/integration/test_bootstrap_interface.py:140]: Please explain why it is (currently) needed.

File: tests/integration/test_bootstrap_interface.py
Comment (rettichschnidi) [inline tests/integration/test_bootstrap_interface.py:10]: What is the reason to have such a wall of text? Looks like something that outdates very fast and is of limited value.

File: tests/integration/test_device_mgmnt_interface.py
Comment (rettichschnidi) [inline tests/integration/test_device_mgmnt_interface.py:16]: Please no commented or dead code.

File: tests/integration/test_device_mgmnt_interface.py
Comment (rettichschnidi) [inline tests/integration/test_device_mgmnt_interface.py:33]: Please no commented or dead code.

File: tests/integration/test_device_mgmnt_interface.py
Comment (rettichschnidi) [inline tests/integration/test_device_mgmnt_interface.py:42]: Please no commented or dead code.

File: tests/integration/test_device_mgmnt_interface.py
Comment (rettichschnidi) [inline tests/integration/test_device_mgmnt_interface.py:37]: Cleanup! ;)

File: tests/integration/test_registration_interface.py
Comment (rettichschnidi) [inline tests/integration/test_registration_interface.py:3]: This link should be moved to a more "generic" place.

File: tests/integration/test_registration_interface.py
Comment (rettichschnidi) [inline tests/integration/test_registration_interface.py:39]: Please no commented or dead code.

File: tests/integration/test_registration_interface.py
Comment (rettichschnidi) [inline tests/integration/test_registration_interface.py:61]: Please no commented or dead code.

File: tests/integration/test_registration_interface.py
Comment (rettichschnidi) [inline tests/integration/test_registration_interface.py:52]: Would remove it until this PR actually got merged. Dead code is bad code.

**Review by rettichschnidi — CHANGES_REQUESTED** [review]
My feedback in general, not specific to a certain file/line of code:
- Please use a linter ([pylint](https://www.pylint.org/) being the obvious choice)
- Please format the code according to PEP8
- Instead of using the example projects, introduce new executable which can be specifically tailored to the testing needs.
- If any TODOs remain in the code, please explain them well

**Comment by qleisan** [comment]
Thanks for the feedback. I'll fix formatting etc. before I push a non-draft version. 
> Instead of using the example projects, introduce new executable which can be specifically tailored to the testing needs.

Interesting thought but do we really want this? Any other feedback on the approach to integration testing (anyone) - any conceptual improvements now might save a lot of test development time

**Comment by rettichschnidi** [comment]
> Interesting thought but do we really want this?

I think it will become increasingly hard have a single target serving two purposes:
* Beeing a simple, easy to understand example
* Supporting everything needed to cover many (all) corner cases for integration testing.

I am pretty sure separate binaries are the way to go.

> Any other feedback on the approach to integration testing (anyone) - any conceptual improvements now might save a lot of test development time

I like the direction (Tests in pytest instead of C, invoking executables, using test cases from the standard).

**Comment by qleisan** [comment]
> I think it will become increasingly hard have a single target serving two purposes:
> 
> - Beeing a simple, easy to understand example
> - Supporting everything needed to cover many (all) corner cases for integration testing.

Today we have a separation in _client_ and _lightclient_. I see your point but not sure that introducing a _testclient_ (and _testserver_, _testbootstrapsever_) is the best way to go, at least not now. Shall we revisit this once it is more clear what modifications are actually needed in order to support good testability?

**Comment by rettichschnidi** [comment]
> I see your point but not sure that introducing a _testclient_ (and _testserver_, _testbootstrapsever_) 

I'd aim for one (configurable) client and one (configurable) server target, not many.

One then can build them multiple times, each time with a different set of supported features and run the relevant pytest tests against it.

> ... is the best way to go, at least not now. Shall we revisit this once it is more clear what modifications are actually needed in order to support good testability?

Feel free to start off with the examples, but when looking at the "Possible improvements" from back then I have the feeling that we will very soon reach a state when dedicated targets, optimized for testing, will be needed.

File: tests/integration/test_bootstrap_interface.py
Comment (rettichschnidi) [reply tests/integration/test_bootstrap_interface.py]: Did you already address those issues?

File: examples/server/lwm2mserver.c
Comment (sbertin-telular) [reply examples/server/lwm2mserver.c:1094]: I think this format is also much harder to read.

**Comment by sbertin-telular** [comment]
I'm glad to see this type of automated testing. Much easier and more reliable to see if something got broken than manual testing.

**Comment by qleisan** [comment]
Pushed a slim down version that can be merged one default block size is set to 1024 (done in PR#596)
Feedback on how to make the `HelperBase` class more pythonic/PEP8 welcome.

File: examples/server/lwm2mserver.c
Comment (qleisan) [reply examples/server/lwm2mserver.c]: The formatting was done by the tooling `git clang-format-10`. The code outdated replaced by PR #602

File: examples/client/CMakeLists.txt
Comment (qleisan) [reply examples/client/CMakeLists.txt:17]: No this shouldn't be here still (belonged to draft)

File: tests/integration/conftest.py
Comment (qleisan) [reply tests/integration/conftest.py:79]: It is the command line arguments you pass. Besides documenting I'm really interested in a nicer way to do it

File: tests/integration/test_registration_interface.py
Comment (qleisan) [reply tests/integration/test_registration_interface.py]: moved to README.md

**Review by tuve — COMMENTED** [review]
I have no issues beyond what has already been commented by @rettichschnidi

File: tests/integration/test_bootstrap_interface.py
Comment (qleisan) [reply tests/integration/test_bootstrap_interface.py]: The integration tests work by parsing the output of the example binaries, this limits what is possible to do. E.g. there is no support to check actual packets. Only information that is printed can be checked etc. What is implemented is what I think is correct or as close as possible (capturing the spirit of the test).

File: tests/integration/conftest.py
Comment (qleisan) [reply tests/integration/conftest.py]: this has been refactored in newer version

File: .github/workflows/build_and_test.yaml
Comment (qleisan) [inline .github/workflows/build_and_test.yaml:51]: Better keep the python version used in the image (so always same)

File: .github/workflows/build_and_test.yaml
Comment (qleisan) [inline .github/workflows/build_and_test.yaml:60]: add 
python -c "import sys; print(sys.version)"

File: .pylintrc
Comment (qleisan) [inline .pylintrc:146]: remove these and update code

Comment (qleisan) [reply .pylintrc]: kept line-too-long

File: .github/workflows/build_and_test.yaml
Comment (qleisan) [reply .github/workflows/build_and_test.yaml:56]: done

File: .github/workflows/build_and_test.yaml
Comment (qleisan) [reply .github/workflows/build_and_test.yaml]: removed

File: tests/integration/conftest.py
Comment (qleisan) [inline tests/integration/conftest.py:72]: suggestions anyone? (2 more instances)

File: .pylintrc
Comment (qleisan) [inline .pylintrc:142]: agree?

**Comment by qleisan** [comment]
@rettichschnidi - please review

**Comment by qleisan** [comment]
Fixed gitlint issue

File: .github/workflows/build_and_test.yaml
Comment (mlasch) [inline .github/workflows/build_and_test.yaml:56]: Either run pytest from the project root with `pytest -v tests/integration` or run pytest directly from within the integration folder. Otherwise pytest.ini will not get evaluated resulting in a warning.

File: tests/integration/test_device_mgmnt_interface.py
Comment (mlasch) [inline tests/integration/test_device_mgmnt_interface.py:26]: This looks like very repetitive code. It's hard to get what should happen here. Some helper function may make the code cleaner to read.

**Review by mlasch — CHANGES_REQUESTED** [review]

**Comment by mlasch** [comment]
The log-files created during the test run should be in the `.gitignore` file. Maybe better even a pattern like `*_log.txt`. Or place them in a separate `runlogs` folder which is then ignored.
All in all this PR looks very promising. I'm looking forward to have these automated tests included.

File: tests/integration/conftest.py
Comment (qleisan) [reply tests/integration/conftest.py:1]: ok!

File: .github/workflows/build_and_test.yaml
Comment (qleisan) [reply .github/workflows/build_and_test.yaml:56]: good catch!

File: tests/integration/test_device_mgmnt_interface.py
Comment (qleisan) [reply tests/integration/test_device_mgmnt_interface.py:26]: ok!

**Comment by qleisan** [comment]
@mlasch pushed update with fixes for review comments (thanks!). Tests running

File: .github/workflows/pylint.yaml
Comment (rettichschnidi) [inline .github/workflows/pylint.yaml:18]: Neither history nor submodules are needed for this.

File: .github/workflows/build_and_test.yaml
Comment (rettichschnidi) [inline .github/workflows/build_and_test.yaml:50]: Why is this needed here, but (apparently) not for gitlint?

File: .github/workflows/pylint.yaml
Comment (rettichschnidi) [inline .github/workflows/pylint.yaml:28]: All of this file should be part of `compliance.yaml`.

File: tests/integration/pytest.ini
Comment (rettichschnidi) [inline tests/integration/pytest.ini:5]: Superfluous newline

File: README.md
Comment (rettichschnidi) [inline README.md:104]: Please add a newline here

File: .pylintrc
Comment (rettichschnidi) [inline .pylintrc:1]: Please either add a minimalistic file (so that we can see what exactly you changed compared to to the default version) or commit the unchanged, vanilla version in a separate commit, then change what you need to be changed in a subsequent commit (again, so that we can see what exactly you changed compared to to the default version).

File: tests/integration/.gitignore
Comment (rettichschnidi) [inline tests/integration/.gitignore:2]: Please add a newline at the end of this line

**Review by rettichschnidi — CHANGES_REQUESTED** [review]
A few things for now, will do a better review (actually looking at the python code) once I can find some more time.

Surely looking much better already!

File: .github/workflows/pylint.yaml
Comment (qleisan) [reply .github/workflows/pylint.yaml:18]: True, will remove

File: .github/workflows/build_and_test.yaml
Comment (qleisan) [reply .github/workflows/build_and_test.yaml:50]: It is better to be consistent, I'll remove the pip upgrade.
Do you think we shall "lock down" all versions of everything we use e.g. python requirements? It makes sense to have fixed versions of everything in a CI flow (same environment for every run). On the other hand this is a small project and it might be less of a hassle to fix version issues when/if they arise than to remember to upgrade versions. We probably don't want the complexity of CI "progressions" testing and updating the versions automatically...

File: .github/workflows/pylint.yaml
Comment (qleisan) [reply .github/workflows/pylint.yaml:28]: ok, I'll move the python linting into this file

File: tests/integration/pytest.ini
Comment (qleisan) [reply tests/integration/pytest.ini:5]: generosity ;-). I'll fix

File: tests/integration/.gitignore
Comment (qleisan) [reply tests/integration/.gitignore:2]: ok

File: README.md
Comment (qleisan) [reply README.md:104]: ok

File: .pylintrc
Comment (qleisan) [reply .pylintrc:1]: Alright

File: tests/integration/pytest.ini
Comment (qleisan) [reply tests/integration/pytest.ini:5]: will remove

File: .pylintrc
Comment (qleisan) [reply .pylintrc:1]: ok

File: README.md
Comment (qleisan) [reply README.md:104]: will also modify to `pytest -v tests/integration` to remove "PytestUnknownMarkWarning"

File: .github/workflows/pylint.yaml
Comment (qleisan) [reply .github/workflows/pylint.yaml:18]: ok will remove

File: .github/workflows/build_and_test.yaml
Comment (rettichschnidi) [reply .github/workflows/build_and_test.yaml:50]: I am not certain, but I'd leave it as it is. Getting to know if newer versions break our builds has some value too.

File: .pylintrc
Comment (rettichschnidi) [reply .pylintrc:1]: One last, tiny nitpick: Would be nice to know which version of pylint you have used to generated the file. Maybe adding this information to the commit message?

File: README.md
Comment (rettichschnidi) [inline README.md:104]: This looks bad when being rendered:
image

Propsal:

```suggestion
```bash
cd wakaama
tools/ci/run_ci.sh --run-build
pytest -v tests/integration
```
```
(you can leave out the "bash", had to add it as otherwise the suggestions functionality breaks.

**Review by rettichschnidi — CHANGES_REQUESTED** [review]
I just can not find the time to review the python code itself. Nevertheless, I am willing to approve after this round.

File: README.md
Comment (qleisan) [reply README.md:104]: Good catch. I added tripple backticks on the line above and below and it renders nicely on my machine...

Comment (qleisan) [reply README.md:104]: added pylint version printout to commit message

**Review by mlasch — APPROVED** [review]

**Review by rettichschnidi — APPROVED** [review]

**Pull request merged**
